src/python/pants/engine/target.py

```diff
@@ -708,8 +708,32 @@ class GenerateTargetsRequest(Generic[_Tgt]):
 class GeneratedTargets(FrozenDict[Address, Target]):
     """A mapping of the address of generated targets to the targets themselves."""
 
-    def __init__(self, generated_targets: Iterable[Target]) -> None:
-        super().__init__({tgt.address: tgt for tgt in generated_targets})
+    def __init__(self, generator: Target, generated_targets: Iterable[Target]) -> None:
+        expected_spec_path = generator.address.spec_path
+        expected_tgt_name = generator.address.target_name
+        mapping = {}
+        for tgt in generated_targets:
+            if tgt.address.spec_path != expected_spec_path:
+                raise ValueError(
```

Comment:
Maybe use `InvalidTargetException` here and the other two raises?

Reply:
Good call.